### PR TITLE
[CostModel][X86] SK_InsertSubvector - if the legalized base/subvector types match then the insert is now just a copy

### PR DESCRIPTION
**Pull request by RKSimon**

### CODE REVIEW

File: llvm/lib/Target/X86/X86TargetTransformInfo.cpp [hunk 1]
```diff
@@ -1550,6 +1550,12 @@ InstructionCost X86TTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
   if (Kind == TTI::SK_InsertSubvector && LT.second.isVector()) {
     int NumElts = LT.second.getVectorNumElements();
     std::pair<InstructionCost, MVT> SubLT = getTypeLegalizationCost(SubTp);
+
+    // If we're inserting into the same legalized type then its not an
+    // insert anymore, its a copy.
+    if (LT.second == SubLT.second)
+      return TTI::TCC_Free;
```
Comment (alexey-bataev) [inline llvm/lib/Target/X86/X86TargetTransformInfo.cpp:1557]: This is probably not quite correct. shufflevector <4 x i32>, <4 xi32> , <0,1,2,4> is a subvector insert of 2 vectors with the same types, but it is not free.

Comment (RKSimon) [reply llvm/lib/Target/X86/X86TargetTransformInfo.cpp:1557]: But wouldn't the SubTp not match?

Comment (RKSimon) [reply llvm/lib/Target/X86/X86TargetTransformInfo.cpp:1557]: Sorry, just realized what you meant - I'll take another look soon.

**Comment by RKSimon** [comment]
broken + defunct

**Pull request closed without merging**